pkg/downloader/downloader.go

```diff
@@ -1,208 +1,6 @@
-		if s, ok := dl.indexLock.Sources[depIdx.AppCode]; ok && s.Source != sourceName {
-			slog.Warn(fmt.Sprintf("%s was already installed from %s.", depIdx.AppCode, s.Source))
-			if !replace {
-				continue
-			}
-			slog.Warn(fmt.Sprintf("Replacing %s with %s.", s.Source, sourceName))
-			delete(dl.indexLock.Bundles, s.Source)
-			replaced[s.Source] = struct{}{}
-		}
-
-		dest := filepath.Join(dl.dependenciesDir, depIdx.AppCode)
-		if _, err := os.Stat(dest); err == nil {
-			if err = os.RemoveAll(dest); err != nil {
-				return nil, nil, err
-			}
-		}
-
-		if _, err = filesys.UnzipToFS(cacheZip, dest); err != nil {
-			return nil, nil, err
-		}
-
-		dl.indexLock.Sources[depIdx.AppCode] = bundle.SourceInfo{
-			Source: sourceName,
-		}
-
-		dl.indexLock.Bundles[sourceName] = bundle.Info{
-			Name:      "",
-			AppCode:   depIdx.AppCode,
-			Integrity: commitHash,
-			Version:   depVersion, // TODO: Use golang pseudo-version format
-			Source:    source,
-			Depends:   depIdx.Depends,
-		}
-
-		if depIdx.Depends != nil {
-			depInstalled, depReplaced, err := dl.Download(depIdx.Depends, replace)
-			if err != nil {
-				return nil, nil, err
-			}
-			installed = append(installed, depInstalled...)
-			for k := range depReplaced {
-				replaced[k] = struct{}{}
-			}
-		}
-
-		installed = append(installed, sourceName)
-	}
-	return installed, replaced, nil
-}
-
-func (dl *goLikeDownloader) loadGitDependency(sourceName string, source string, ref string, hash string) (string, error) {
-	filename := fmt.Sprintf("%s-%s-%s.zip", filepath.Base(sourceName), ref, hash[:8])
-	cacheZip := filepath.Join(dl.cacheDir, filepath.Dir(sourceName), filename)
-	// If cached ZIP does not exist - fetch the archive
-	if _, err := os.Stat(cacheZip); err != nil {
-		if err = os.MkdirAll(filepath.Join(dl.cacheDir, filepath.Dir(sourceName)), 0755); err != nil {
-			return "", err
-		}
-		// TODO: Ref discovery
-		slog.Info(fmt.Sprintf("Cache miss. Loading from: %s", source))
-		if err = gitArchive(source, ref, cacheZip); err != nil {
-			return "", err
-		}
-	} else {
-		slog.Info(fmt.Sprintf("Cache hit. Loading %s from cache.", filename))
-	}
-	return cacheZip, nil
-}
-
-func (dl *goLikeDownloader) discoverSource(source string) ([]byte, error) {
-	// TODO: Better dependency path handling
-	// Reuse the same resolution mechanism that go mod uses
-	// https://go.dev/ref/mod#vcs-find
-	url, err := url.Parse(source)
-	if err != nil {
-		return nil, err
-	}
-	query := url.Query()
-	query.Add("go-get", "1")
-
-	resp, err := http.Get(url.String() + "?" + query.Encode())
-	if err != nil {
-		return nil, err
-	}
-	defer resp.Body.Close()
-
-	return io.ReadAll(resp.Body)
-}
-
-func (dl *goLikeDownloader) parseGoQuery(goQuery string) (string, string, string) {
-	parts := strings.Split(goQuery, " ")
-	return parts[0], parts[1], parts[2]
+	Discover(string, string) (Downloader, string, error)
```

Comment:
I think `Discover` is implementation-specific. Maybe use generics here if you want stricter interface?